### PR TITLE
start of developing server side for options

### PR DESCRIPTION
**Pull request by cle1000**
#not finished yet# 
i added just a new route with a request handler for options

### CODE REVIEW

File: web/src/js/components/footer.js [hunk 1]
```diff
@@ -15,6 +15,26 @@ export default function Footer({ settings }) {
             {intercept && (
                 <span className="label label-success">Intercept: {intercept}</span>
             )}
+            {showhost && (
+                <span className="label label-success">showhost</span>
+            )}
+            {no_upstream_cert && (
+                <span className="label label-success">no-upstream-cert</span>
+            )}
+             {rawtcp && rawtcp != "False" && (
```
Comment (mhils) [inline web/src/js/components/footer.js]: `!= "False"`

File: mitmproxy/web/app.py [hunk 2]
```diff
@@ -286,7 +286,13 @@ def get(self):
...
                 version=version.VERSION,
                 mode=str(self.master.server.config.mode),
-                intercept=self.state.intercept_txt
+                intercept=self.state.intercept_txt,
+                showhost=self.master.options.showhost,
+                no_upstream_cert=self.master.server.config.no_upstream_cert,
+                rawtcp=self.master.server.config.rawtcp,
+                http2=self.master.server.config.http2,
+                anticache=self.master.options.anticache,
+                anticomp=self.master.options.anticomp
```
Comment (mhils) [inline mitmproxy/web/app.py]: `self.master.anticache` directly

File: mitmproxy/web/app.py [hunk 3]
```diff
@@ -296,6 +302,24 @@ def put(self):
...
+                self.master.server.config.rawtcp = v
+                update[k] = v
+            elif k == "http2":
+                self.master.server.config.http2 = v
+                update[k] = v
+            elif k == "anticache":
+                self.master.anticache = v
+                update[k] = v
+            elif k == "anticomp":
+                self.master.anticomp = v
```
Comment (mhils) [inline mitmproxy/web/app.py]: `self.master.options.anticomp`

**Comment by mhils** [comment]
I'll fix the settings inconsistency.

File: web/src/js/components/header.js [hunk 4]
```diff
@@ -250,43 +251,41 @@ var ViewMenu = React.createClass({
...
-        ]
-      }
-    }
-    setOption(entry){
-      console.log(entry.name);//TODO: get options from outside and remove state
-      entry.checked = !entry.checked;
-      this.setState({options: this.state.options});
+// If I make a stateless function out of the option menu i got the following warning message:
+// Warning: Stateless function components cannot be given refs (See ref "active" in OptionMenu created by Header). Attempts to access this ref will fail.
+class OptionMenu extends React.Component {
```
Comment (mhils) [inline web/src/js/components/header.js]: Let's remove ref="active" in `Header` and make this stateless again. :smiley:

File: web/src/js/components/header.js [hunk 5]
```diff
@@ -250,43 +251,41 @@ var ViewMenu = React.createClass({
...
+
+    onOptionChange(name, checked) {
+        SettingsActions.update(JSON.parse("{\"" + name + "\": " + !checked + "}"));
```
Comment (gzzhanghao) [inline web/src/js/components/header.js]: Maybe `SettingsActions.update({ [name]: !checked + '' })` ?

Comment (cle1000) [inline web/src/js/components/header.js]: thanks, i'll try :+1:

File: mitmproxy/web/app.py [hunk 6]
```diff
@@ -315,10 +318,19 @@ def put(self):
...
+                self.master.options.anticomp = v
+                update[k] = v
+            elif k == "stickycookie":
+                self.master.set_stickycookie(v)
+                update[k] = v
+            elif k == "stickyauth":
+                self.master.set_stickyauth(v)
+                update[k] = v
+            elif k == "stream":
+                #self.master.set_stream_large_bodies(v) todo: ask
```
Comment (cle1000) [inline mitmproxy/web/app.py]: can anybody tell me how to set stream, should i set 
-- self.master.stream 
-- self.master.stream_large_bodies
-- ...

thanks

File: web/src/js/components/common.js [hunk 7]
```diff
@@ -133,14 +135,53 @@ export var Splitter = React.createClass({
...
+        <span className={"fa " + (props.checked ? "fa-check-square-o" : "fa-square-o")}>&nbsp;{props.name}</span>
+        </div>
+    </div>;
+
+ToggleButton.propTypes = {
+    name: React.PropTypes.string.isRequired,
+    onToggleChanged: React.PropTypes.func.isRequired
+};
+
+export class ToggleInputButton extends React.Component {
```
Comment (cle1000) [inline web/src/js/components/common.js]: should i also add a helper field, like for filters ?

File: mitmproxy/web/app.py (hunk 6 above)
Comment (mhils) [inline mitmproxy/web/app.py]: You can call `FlowMaster.set_stream_large_bodies` :smiley:

File: web/src/js/components/common.js (hunk 7 above)
Comment (mhils) [inline web/src/js/components/common.js]: I think this is fine as-is. We may want to add a help thing for this later, but we have plenty of stuff that is probably more important right now.

**Comment by cle1000** [comment]
i think this options is ready to merge, what do you think?

**Comment by mhils** [comment]
:+1: 

I'm not completely satisfied with the UX of ToggleInputButton yet, but these improvements don't have to be a part of this PR. For example, it's weird that you can't click the checkbox when the input is empty. Ideally, empty input + checkbox = apply to all.

Anyway, we're really making progress here! :smiley: :cake: :tada:

**Pull request merged**